homeassistant/components/roomba/vacuum.py

```diff
@@ -83,57 +54,52 @@
 SUPPORT_ROOMBA_CARPET_BOOST = SUPPORT_ROOMBA | SUPPORT_FAN_SPEED
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the iRobot Roomba vacuum cleaner platform."""
-
-    if PLATFORM not in hass.data:
-        hass.data[PLATFORM] = {}
-
-    host = config.get(CONF_HOST)
-    name = config.get(CONF_NAME)
-    username = config.get(CONF_USERNAME)
-    password = config.get(CONF_PASSWORD)
-    certificate = config.get(CONF_CERT)
-    continuous = config.get(CONF_CONTINUOUS)
-    delay = config.get(CONF_DELAY)
-
-    roomba = Roomba(
-        address=host,
-        blid=username,
-        password=password,
-        cert_name=certificate,
-        continuous=continuous,
-        delay=delay,
-    )
-    _LOGGER.debug("Initializing communication with host %s", host)
-
-    try:
-        with async_timeout.timeout(9):
-            await hass.async_add_job(roomba.connect)
-    except asyncio.TimeoutError:
-        raise PlatformNotReady
-
-    roomba_vac = RoombaVacuum(name, roomba)
-    hass.data[PLATFORM][host] = roomba_vac
-
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the iRobot Roomba vacuum cleaner."""
+    roomba = hass.data[DOMAIN][config_entry.entry_id]
+    roomba_vac = RoombaVacuum(roomba)
     async_add_entities([roomba_vac], True)
 
 
 class RoombaVacuum(VacuumDevice):
     """Representation of a Roomba Vacuum cleaner robot."""
 
-    def __init__(self, name, roomba):
+    def __init__(self, roomba):
         """Initialize the Roomba handler."""
         self._available = False
         self._battery_level = None
         self._capabilities = {}
         self._fan_speed = None
         self._is_on = False
-        self._name = name
         self._state_attrs = {}
         self._status = None
         self.vacuum = roomba
-        self.vacuum_state = None
+        self.vacuum_state = self.vacuum.master_state.get("state", {}).get(
+            "reported", {}
+        )
+        self._mac = self.vacuum_state.get("mac")
+        self._name = self.vacuum_state.get("name")
+
+    @property
+    def unique_id(self):
+        """Return the uniqueid of the vacuum cleaner."""
+        return "roomba_{}".format(self._mac)
+
+    @property
+    def device_info(self):
+        """Return the device info of the vacuum cleaner."""
+        return {
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "connections": {(CONNECTION_NETWORK_MAC, self._mac)},
+            "manufacturer": "iRobots",
+            "name": str(self._name),
+            "sw_version": self.vacuum.master_state.get("state", {})
```

Comment:
Looks like these can be switched to `self.vacuum_state`

If they are ever going to be empty then they should not be added to the dict()